Define an object for collecting DSLX interpreter events.

meheff opened the pull request:
The newly defined DslxInterpreterEvents gathers assert and trace messages from the DSLX interpreter replacing the previous trace hook mechanism. It stores information in the same proto as the IR interpreter does which enables easier comparison of behavior between the interpreters and the sharing of tools. Also augment the proto with additional information like source information.

Also add a utility dump_call_trace which reads an events proto and prints a structured trace of the calls.

proppy commented on xls/dslx/interpreter_main.cc:
```diff
@@ -206,6 +210,12 @@ absl::StatusOr<TestResult> RealMain(
   };
 
+  // Create a results proto if requested and plumb it through options.
+  xls::EvaluatorResultsProto results_proto;
```
yay structured logs \o/

meheff commented on xls/ir/evaluator_result.proto:
```diff
@@ -30,6 +36,13 @@ message TraceCallProto {
 
+// Message metadata for a function return when call tracing is enabled.
+message TraceCallReturnProto {
+  optional string function_name = 1;
```
There is a one-to-one correspondence between TraceCallProto and TraceCallReturnProto. The trace arguments are recorded with the TraceCallProto. This is done this way because the values are recorded when they are immediately available (generally in an evaluator you may not simultaneously have access to the arguments and return value because they are computed as separate times). Extended the comment to clarify this.

meheff commented:
Any progress on this? Thanks.

ericastor commented:
@meheff Please rebase this when you can; we can't pull it in for final review until conflicts are resolved.

meheff commented:
Rebased to head... and changed my notifications for the XLS repo to my personal email account as I kept missing responses on my PRs (my work email might as well be /dev/null).

mikex-oss commented on xls/dev_tools/BUILD:
```diff
@@ -924,3 +924,33 @@ pytype_strict_library(
+    ],
+)
+
+py_test(
```
For future reference, please ensure you load `py_test` from rules_python, as the builtin is deprecated.

You can also use https://github.com/google/xls/blob/main/xls/build_rules/py_oss_defs.bzl, which might make sense here, since you added type annotations in the test file.

mikex-oss commented on xls/dev_tools/dump_call_trace_test.py:
```diff
@@ -0,0 +1,259 @@
+from typing import List
+
+from absl.testing import absltest
+from textwrap import dedent
```
FYI, this causes an error in the internal linter. Don't import functions.

http://google.github.io/styleguide/pyguide.html#22-imports

mikex-oss commented on xls/dev_tools/dump_call_trace_test.py:
```diff
@@ -0,0 +1,259 @@
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+from __future__ import annotations
```
This appears to be unused.

mikex-oss reviewed (commented):
@meheff I tried patching this internally (most of the comments are just FYI), but I don't actually know why the tests have different start indices. Any ideas?

proppy commented on xls/dev_tools/dump_call_trace_test.py:
```diff
@@ -0,0 +1,259 @@
+    out = _run_dump(path)
+    dslx = self._proto_path('prog.x')
+    expected = dedent(f"""
+    {dslx}:6:0: call_trace_test(
```
This is also failing externally, maybe we can exclude the line number from the match?

proppy commented on xls/dev_tools/dump_call_trace_test.py:
```diff
@@ -0,0 +1,259 @@
+      token
+      bits[1]:1
+    )
+      {dslx}:5:13: ____itok__prog__for_loop_inside_test_counted_for_0_body(
```
similarly here, can we exclude the line number from the match?

proppy commented on xls/dev_tools/dump_call_trace_test.py:
```diff
@@ -0,0 +1,259 @@
+          bits[32]:1
+        )
+        __prog__foo(...) => bits[32]:2
+        {dslx}:7:15: __prog__foo(
```
similarly here, can we exclude the line number from the match?

proppy commented on xls/dev_tools/dump_call_trace_test.py:
```diff
@@ -0,0 +1,259 @@
+        )
+        __prog__foo(...) => bits[32]:12
+      ____itok__prog__for_loop_inside_test_counted_for_0_body(...) => (token, bits[1]:1, bits[32]:0)
+      {dslx}:5:13: ____itok__prog__for_loop_inside_test_counted_for_0_body(
```
similarly here, can we exclude the line number from the match?

proppy commented on xls/dev_tools/dump_call_trace_test.py:
```diff
@@ -0,0 +1,259 @@
+          bits[32]:0
+        )
+      )
+        {dslx}:6:15: __prog__foo(
```
similarly here, can we exclude the line number from the match?

proppy commented on xls/dev_tools/dump_call_trace_test.py:
```diff
@@ -0,0 +1,259 @@
+          bits[32]:2
+        )
+        __prog__foo(...) => bits[32]:3
+        {dslx}:7:15: __prog__foo(
```
similarly here, can we exclude the line number from the match?

proppy reviewed (changes requested):
Can we relax the check in `dump_call_trace_test` to exclude the line numbers from the match?

proppy commented:
this also need to be rebased.

meheff commented on xls/dev_tools/dump_call_trace_test.py:
```diff
@@ -0,0 +1,259 @@
+    out = _run_dump(out_path)
+    dslx = self._proto_path('prog.x')
+    expected = dedent(f"""
+    {dslx}:4:0: for_loop_inside_test(
```
Fixed.

proppy commented:
@meheff sorry for the delay, can you rebase again?
